engine/consensus/provider/engine_test.go

```diff
@@ -17,142 +17,147 @@ import (
+type Suite struct {
+	suite.Suite
 
-	me := &module.Local{}
-	state := &protocol.State{}
-	final := &protocol.Snapshot{}
-	con := &network.Conduit{}
-	metrics := metrics.NewNoopCollector()
+	me         *module.Local
+	conduit    *network.Conduit
+	state      *protocol.State
+	final      *protocol.Snapshot
+	epochQuery *protocol.EpochQuery
+	nextEpoch  *protocol.Epoch
 
-	e := &Engine{
-		me:      me,
-		state:   state,
-		con:     con,
-		message: metrics,
-		tracer:  trace.NewNoopTracer(),
-	}
-
-	proposal := unittest.ProposalFixture()
-	identities := unittest.IdentityListFixture(100)
-	localID := identities[0].NodeID
-
-	params := []interface{}{proposal}
-	for _, id := range identities[1:] {
-		params = append(params, id.NodeID)
-	}
+	currentEpochIdentities flow.IdentityList
+	nextEpochIdentities    flow.IdentityList
 
-	me.On("NodeID").Return(localID)
-	state.On("Final").Return(final).Once()
-	final.On("Identities", mock.Anything).Return(identities[1:], nil).Once()
-	con.On("Publish", params...).Return(nil).Once()
-
-	err := e.onBlockProposal(localID, proposal)
-	require.NoError(t, err)
+	engine *Engine
+}
 
-	me.AssertExpectations(t)
-	state.AssertExpectations(t)
-	final.AssertExpectations(t)
-	con.AssertExpectations(t)
+func TestProviderEngine(t *testing.T) {
+	suite.Run(t, new(Suite))
 }
 
-func TestOnBlockProposalRemoteOrigin(t *testing.T) {
+func (suite *Suite) SetupTest() {
 
-	me := &module.Local{}
-	state := &protocol.State{}
-	final := &protocol.Snapshot{}
-	con := &network.Conduit{}
-	metrics := metrics.NewNoopCollector()
+	suite.me = new(module.Local)
+	suite.conduit = new(network.Conduit)
+	suite.state = new(protocol.State)
+	suite.final = new(protocol.Snapshot)
+	suite.epochQuery = new(protocol.EpochQuery)
+	suite.nextEpoch = new(protocol.Epoch)
 
-	e := &Engine{
-		me:      me,
-		state:   state,
-		con:     con,
-		message: metrics,
+	suite.engine = &Engine{
+		me:      suite.me,
+		state:   suite.state,
+		con:     suite.conduit,
+		message: metrics.NewNoopCollector(),
 		tracer:  trace.NewNoopTracer(),
 	}
-	proposal := unittest.ProposalFixture()
-	identities := unittest.IdentityListFixture(100)
-	localID := identities[0].NodeID
-	remoteID := identities[1].NodeID
-
-	me.On("NodeID").Return(localID)
 
-	err := e.onBlockProposal(remoteID, proposal)
-	require.Error(t, err)
+	suite.currentEpochIdentities = unittest.CompleteIdentitySet()
+	suite.nextEpochIdentities = unittest.CompleteIdentitySet()
+	localID := suite.currentEpochIdentities[0].NodeID
+
+	suite.me.On("NodeID").Return(localID)
+	suite.state.On("Final").Return(suite.final)
+	suite.final.On("Identities", mock.Anything).Return(
+		func(f flow.IdentityFilter) flow.IdentityList { return suite.currentEpochIdentities[1:].Filter(f) },
+		func(flow.IdentityFilter) error { return nil },
+	)
+	suite.final.On("Epochs").Return(suite.epochQuery)
+	suite.epochQuery.On("Next").Return(suite.nextEpoch)
+	suite.nextEpoch.On("InitialIdentities").Return(
```

Comment:
I guess here we can just do `.Return(suite.nextEpochIdentities, nil)`

Reply:
I try to use the functional returns in the mocks when it's returning members of `Suite`, because it will stay up to date with the current value of `suite.nextEpochIdentities` if it's changed. If you do `.Return(suite.nextEpochIdentities, nil)`, it will always return the initial value of `suite.nextEpochIdentities` since `suite.nextEpochIdentities` is evaluated when `Return` is invoked in `SetupTest`.

Reply:
that makes a lot of sense! thanks